File: src/python/tests/intervals.py
# Copyright (c) 2015-2017 by the parties listed in the AUTHORS file.
# All rights reserved.  Use of this source code is governed by
# a BSD-style license that can be found in the LICENSE file.

from ..mpi import MPI
from .mpi import MPITestCase

import sys
import os

import numpy as np
import numpy.testing as nt

from ..dist import *
from ..tod.interval import *
from ..tod.sim_interval import *


class IntervalTest(MPITestCase):

    def setUp(self):
        self.rate = 123.456
        self.duration = 24 * 3601.23
        self.gap = 3600.0
        self.start = 5432.1
        self.first = 10
        self.nint = 3


    def test_regular(self):
        intrvls = regular_intervals(self.nint, self.start, self.first,
                                    self.rate, self.duration, self.gap)

        goodsamp = self.nint * ( int(self.duration * self.rate) + 1 )

        check = 0

        for it in intrvls:
            #print(it.first," ",it.last," ",it.start," ",it.stop)
            check += it.last - it.first + 1

        nt.assert_equal(check, goodsamp)
        return

File: src/toast/_libtoast/intervals.cpp
// Copyright (c) 2015-2021 by the parties listed in the AUTHORS file.
// All rights reserved.  Use of this source code is governed by
// a BSD-style license that can be found in the LICENSE file.

#include <module.hpp>

#include <intervals.hpp>


void init_intervals(py::module & m) {
    py::class_ <Interval> (
        m, "Interval",
        R"(
        Numpy dtype for an interval
        )")
    .def(py::init([]() {
                      return Interval();
                  }))
    .def_readwrite("start", &Interval::start)
    .def_readwrite("stop", &Interval::stop)
    .def_readwrite("first", &Interval::first)
    .def_readwrite("last", &Interval::last)
    .def("astuple",
         [](const Interval & self) {
             return py::make_tuple(self.start, self.stop, self.first, self.last);
         })
    .def_static("fromtuple", [](const py::tuple & tup) {
                    if (py::len(tup) != 4) {
                        throw py::cast_error("Invalid size");
                    }
                    return Interval{tup[0].cast <double>(),
                                    tup[1].cast <double>(),
                                    tup[2].cast <int64_t>(),
                                    tup[3].cast <int64_t>()};
                });


    PYBIND11_NUMPY_DTYPE(Interval, start, stop, first, last);
}
